### PR TITLE
Extension serve combined progress bar for host theme creation and sync

### LINKED ISSUE #2524: [Enhancement] TAE: Use the progress bar when the host theme gets created
When users don't have a host theme, the CLI auto creates one:

image

The CLI presents a progress bar when it is cloning the default theme, and it shows many "Synced" when it's pushing it. 

The goal of this task is showing a single progress bar for the host theme creation.

### PR DESCRIPTION
**Pull request by mgmanzella**
### WHY are these changes introduced?

Fixes #2524 

### WHAT is this pull request doing?

Creates new class `HostThemeProgressBar` to output combined progress bar for cloning host theme and syncing files 

#### Single progress bar for `extension serve` 
- For setup, spin can be used with the steps in [this PR](https://github.com/Shopify/storefront-renderer/pull/14290). OR in production, enable the beta flag `theme_extensions_preview_changes` for your store and use `shopify-dev login`
- In `lib/shopify_cli/theme/extension/host_theme.rb` replace `ensure_exists!` with the following code:
```
 def ensure_exists!
      create
      @ctx.debug("Created temporary host theme: #{@id}")
    self
  end
```
- In a terminal run `shopify-dev extension serve`
- Confirmed single progress bar is output and dev server starts

Screen Shot 2022-08-15 at 1 35 25 PM

#### Regression test `theme serve`
- In `lib/shopify_cli/theme/development_theme.rb` replace `ensure_exists!` with the following code:
```
def ensure_exists!
    create
    @ctx.debug("Created temporary development theme: #{@id}")
    ShopifyCLI::DB.set(development_theme_id: @id)

  self
end
```
- If you haven't logged into your store, run `shopify-dev login`
- In a terminal run `shopify-dev theme serve`
- Confirm progress bar and file syncing output hasnt changed

Screen Shot 2022-08-15 at 1 29 18 PM

### Open Questions
- Do we want to also create a combined progress bar for `theme serve`? (Out of the context of extension serve of course, if so I can create an issue)

### CODE REVIEW

File: test/shopify-cli/git_test.rb [hunk 1]
```diff
@@ -195,6 +225,70 @@ def test_clones_git_repo_when_directory_exists_but_it_is_empty
...
+    def test_clone_progress_parses_percentage_and_yield_to_block
+      msg = nil
+
+      clone_msg_stream = [
+        "Receiving objects: 7% (5824/5824), 8.78 MiB | 19.86 MiB/s, done.",
+        "Receiving objects: 91% (5824/5824), 8.78 MiB | 19.86 MiB/s, done.",
+        "Receiving objects: 100% (5824/5824), 8.78 MiB | 19.86 MiB/s, done.",
+      ]
+      ShopifyCLI::Git.clone_progress(stderr(clone_msg_stream), bar: nil) do |percent|
+        msg, _err = capture_io do
```
Comment (jessetuglu) [inline test/shopify-cli/git_test.rb:258]: Using `capture_io` 😄!

File: lib/shopify_cli/git.rb [hunk 2]
```diff
@@ -197,6 +271,36 @@ def sparse_checkout(repo, set, branch, ctx)
...
+          next unless line.strip.start_with?("Receiving objects:")
+          percent = (line.match(/Receiving objects:\s+(\d+)/)[1].to_f / 100).round(2)
+
+          if block_given?
+            yield percent
+          elsif !bar.nil?
+            bar.tick(set_percent: percent)
+          end
+
+          next
```
Comment (jessetuglu) [inline lib/shopify_cli/git.rb:298]: is this `next` required?

File: lib/shopify_cli/theme/extension/ui/host_theme_progress_bar.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+          end
+
+          def progress(method, **args)
+            @syncer.lock_io!
+            CLI::UI::Progress.progress do |bar|
+              Git.public_send(:raw_clone, "https://github.com/Shopify/dawn.git", @dir) do |percent|
+                bar.tick(set_percent: percent * GIT_CLONE_PROGRESS_SHARE)
+              end
+
+              @syncer.public_send(method, **args) do |left, total|
```
Comment (jessetuglu) [inline lib/shopify_cli/theme/extension/ui/host_theme_progress_bar.rb:22]: I assume total can never be `0` with the way `Syncer#wait!` is implemented

**Review by ulgut — COMMENTED** [review]
Just some minor comments/questions, but testing it looked really good! 🚀 Thank you 🙏

File: lib/shopify_cli/git.rb [hunk 4]
```diff
@@ -46,12 +46,75 @@ def sha(dir: Dir.pwd, ctx: Context.new)
...
+      # #### Example
+      #
+      #   ShopifyCLI::Git.raw_clone('git@github.com:shopify/test.git', 'test-app')
+      #
+      def raw_clone(repo_with_branch, dest, ctx: Context.new)
+        if Dir.exist?(dest) && !Dir.empty?(dest)
+          ctx.abort(ctx.message("core.git.error.directory_exists"))
+        else
+          msg = []
+          require "open3"
```
Comment (jessetuglu) [inline lib/shopify_cli/git.rb:96]: I don't think that if `raw_clone` and `clone` are both called `open3` will be required twice?

**Review by ulgut — APPROVED** [review]
🚀

**Review by karreiro — APPROVED** [review]
Awesome stuff 🚀 Thank you, @mgmanzella!

**Pull request merged**